Set elastic password and generate enrollment token

BigPandaToo opened the pull request:
This change implements a class which can be invoked
as a CLI tool and generate an elastic user password
(if `bootstrap.password` doesn't exist) and Kibana
enrollment token for initial node.
If `bootstrap.password` presents, no attempts to
auto-config the elastic password or to "promote"
this password value automatically to the security index
is made.
The process output password, Kibana enrollment token
and fingerprint of the CA certificate that ES is configured
to use for the HTTP layer to standard output.
This functionality is intended for the archive installation.
 



Resolves #75310

BigPandaToo committed:
Set elastic password and generate enrollment token

Set elastic password and generate enrollment token for initial node

BigPandaToo committed:
Set elastic password and generate enrollment token

Set elastic password and generate enrollment token for initial node

BigPandaToo committed:
Set elastic password and generate enrollment token

Set elastic password and generate enrollment token for initial node

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+    private static final Setting<SecureString> BOOTSTRAP_ELASTIC_PASSWORD = SecureSetting.secureString("bootstrap.password",
+        null);
+    private static final Setting<SecureString> CREDENTIALS_PASSWORD = SecureSetting.secureString("bootstrap.password",
+        SEED_SETTING);
```
These shouldn't be redefined like this, the existing ones should be referenced here as well.

In addition, what's the plan with CREDENTIALS_PASSWORDS? It's not clear to me.

BigPandaToo replied:
You are right about redefinition, change that.
However, BOOTSTRAP_ELASTIC_PASSWORD in ReservedRealm defined the way it will be backed up by the keystore.seed. I need to look at the bootstrap.password to see if it was provided,  so I have to define a new variable here to read from the keystore.
Credentials password is what BOOTSTRAP_ELASTIC_PASSWORD is in the reserved realm, it will be used to authenticate while sending requests to the node. I created a new variable here to avoid confusion.

albertzaharovits replied:
I think:
```
if (env.settings().hasValue(ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD.getKey())) {
```
does what you're after, no?

BigPandaToo committed:
Remove redefinition

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+        terminal.println("CA fingerprint: " + fingerprint);
+    }
+
+    protected void checkClusterHealth(Environment env, CommandLineHttpClient client) throws Exception {
```
Please merge master and factor out this method, because it used in other places. I would suggest we add a method to the `CommandLineHttpClient`, but I would be satisfied if we avoid coding this everytime we need it.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+            throw new UserException(ExitCodes.NOOP, "Enrollment is explicitly disabled.");
+        }
+        CommandLineHttpClient client = getClient(env);
+        CreateEnrollmentToken cet = getCreateEnrollmentToken(env);
```
nit: this is not how we name variables in general.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/tool/SetupPasswordTool.java:
```diff
@@ -564,7 +564,7 @@ private URL createURL(URL url, String path, String query) throws MalformedURLExc
     }
 
-    private String getErrorCause(HttpResponse httpResponse) {
+    public static String getErrorCause(HttpResponse httpResponse) {
```
Lets also move this to `CommandLineHttpClient`.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
```
I like this! I vaguely remember Ioannis describing something like this.

Lets make it more visible, ie a public static variable to this class, or maybe an utils class for the enrollment package, as we generate passwords in a couple of places already.
I think there's other code as well that can be reused, eg `generatePassword` .

albertzaharovits replied:
I understand that since we're not going to create the bash script to call this, we don't have to code such a parameter.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
+            throw new UserException(ExitCodes.NOOP, "This command is not intended for end users");
```
My thoughts when introducing this error code were for it to be used when requesting the creation of something that already exists. For the initial Security autoconfiguration the idea was that if Security is already configured then the autoconfiguration is a noop.

I would pick `ExistCodes.CONFIG` for this particular case.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
+            throw new UserException(ExitCodes.NOOP, "This command is not intended for end users");
+        }
+        if (env.settings().hasValue(XPackSettings.ENROLLMENT_ENABLED.getKey()) && false ==
+            XPackSettings.ENROLLMENT_ENABLED.get(env.settings())) {
+            throw new UserException(ExitCodes.NOOP, "Enrollment is explicitly disabled.");
+        }
```
I think this ought not run if enrollment is disabled in any way.

BigPandaToo replied:
It shouldn't. But it is still possible to invoke it. Even though the enrollment token won't be generated, but attempt to password reset can happen

albertzaharovits replied:
I mean that the check doesn't have to be for _explicitly_ disabled. Any disabled is the same error. There is no explicit disabled enrollment anyway.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+        } else {
+            password = bootstrapPassword.get(env.settings());
+        }
+        token = cet.createKibanaEnrollmentToken(elasticUser, password);
+        Map<String, String> infoNode = getDecoded(token);
+        fingerprint = infoNode.get("fgr");
```
Encoding then decoding the token to access the fields is a sign we have the wrong abstraction.
I suggest the enrollment token be a class with an `encode` method, as well as other methods to expose the information inside.

BigPandaToo replied:
Agree about exposing information through the class. Not sure why would we need encode though

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java:
```diff
@@ -0,0 +1,239 @@
+                Thread.sleep(1000);
+            }
+        }
+        if (Strings.isNullOrEmpty(bootstrapPassword.get(env.settings()).toString())) {
+            changeElasticUserPassword(env, terminal, client);
+        } else {
+            password = bootstrapPassword.get(env.settings());
+        }
```
IMO this sort of high level design condition is unsuitable at this level of abstraction in the code. We are likely to introduce other checks like the one for the password hash set by the package installation script, and we should maybe reuse this cmd line utility for generating enrollment tokens apart from the first one.

I think this tool should receive by parameter whether it should reset the password in addition to generating the enrollment token (later we can add which types of enrollment tokens to generate).
If the arguments are incompatible with the configuration it should error out.

Lets discuss this sync.

BigPandaToo replied:
We will discuss it in our sync. Current version doesn't address it

albertzaharovits replied:
We decided this is indeed the good path.

albertzaharovits reviewed (commented):
I did a first pass.

In addition to the things mentioned, I think you need to consider that the node's keystore is not open, prompt for password etc.

Let's discuss sync.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Pwd_initNode

albertzaharovits commented:
We've considered whether this tool is redundant wrt the other tools we've got for resetting the elastic user's password and generating enrollment tokens.
It is not, because:
* the others expect the node to be running, whereas this one has to wait for the node to come up
* this one has to operate only once during node bootstrap
* because it both sets the password and generates enrollment tokens it takes away from the bash code that would otherwise have to call two other cmd line tools; internally this already reuses code from `CreateEnrollmentToken`, and it is always preferable to have java glue code to bash glue code.

albertzaharovits commented:
WRT to having a cmd line tool that does one thing (generates enrollment tokens) but which also "sometimes" does another thing (resets the elastic password unless `bootstrap.password` is set), we're acknowledging this is unfortunate, but this code is going to be invoked by scripts in exactly this configuration every time. No one is going to call this to only generate enrollment tokens or only to reset the passwords.
To minimize surprising anyone, we're going to rename this to something more vague but more illustrative  of the context this is invoked in, ie something containing "Bootstrap" , eg "BootstrapCredentialsInitialNodeTool".

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Pwd_initNode

BigPandaToo committed:
- Moving code under the tool
- Refactoring to reuse some code
- Refactoring checkClusterHealth code
- Change to decrypt keystore
- Update tests

BigPandaToo committed:
Addressing PR feedback

BigPandaToo committed:
Addressing PR feedback

BigPandaToo committed:
Addressing PR feedback

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,195 @@
+        keyStoreWrapper = getKeyStoreWrapper(env);
+        ReadBootstrapPassword(env, terminal);
+        checkClusterHealthWithRetries(client, terminal, elasticUser, credentialsPassword, 0, 5, false);
+        if (Strings.isNullOrEmpty(bootstrapPassword.toString())) {
```
Let's discuss if we need to decide it here or if the caller should make this decision

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseClientAwareCommand.java:
```diff
@@ -0,0 +1,150 @@
+     * newly created file realm user.
+     */
+    protected void checkClusterHealthWithRetries(CommandLineHttpClient client, Terminal terminal, String username, SecureString password,
+                                                 int retriesWaitingForUser, int retriesWaitingForNode, boolean force) throws Exception {
```
Two types of retries introduced here. They can be used together or separately, depending on the caller logic. Waiting for the user will only retry if authentication fails. Waiting for the node will retry if connection is refused or node is Red (making N retries all together, for ex. try to time until connection is established and 3 more if the cluster is Red after it)

BigPandaToo committed:
Addressing PR feedback

BigPandaToo commented:
@albertzaharovits @jkakavas I finished the refactoring. Let's discussed open qq not addressed in this version in the sync

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Pwd_initNode

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+        keyStoreWrapper = getKeyStoreWrapper(env);
+        ReadBootstrapPassword(env, terminal);
+        checkClusterHealthWithRetries(client, terminal, elasticUser, credentialsPassword, 0, 5, false);
+        if (Strings.isNullOrEmpty(bootstrapPassword.toString())) {
```
Because bootstrap password defaults to keystore_seed, this will _never_ be empty. You need to check if bootstrap.password exists in the secure settings instead

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+        }
+    }
+
+    protected void changeElasticUserPassword(Terminal terminal, CommandLineHttpClient client) throws Exception {
```
let's call this `setElasticUserPassword`

jkakavas replied:
can we have this return the password instead of being void and setting a value to a class var ?

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+            if (response.getHttpStatus() != HttpURLConnection.HTTP_OK) {
+                terminal.errorPrintln("");
+                terminal.errorPrintln(
+                    "Unexpected response code [" + response.getHttpStatus() + "] from calling PUT " + passwordChangeUrl.toString());
```
no need to explicitlly call `toString()` on URL

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+            }
+        } catch (IOException e) {
+            terminal.errorPrintln("");
+            terminal.errorPrintln("Connection failure to: " + passwordChangeUrl.toString() + " failed: " + e.getMessage());
```
no need to explicitlly call `toString()` on URL

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+                    return Strings.toString(xContentBuilder);
+                }, this::responseBuilder);
+            if (response.getHttpStatus() != HttpURLConnection.HTTP_OK) {
+                terminal.errorPrintln("");
```
@albertzaharovits would we be capturing stdout/stderr and piping it somewhere when we spawn this process from the startup script? i.e. do we need to handle output ? If not let's remove all these, otherwise we need to be more specific as to what we're printing so that it makes sense to anyone reading it (users are not aware of our implementation details so some context would be helpful  - what is this password change, who invoked it etc )

jkakavas replied:
We decided to not print to output in this class in our last sync right?

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+        fingerprint = createEnrollmentToken.getFingerprint();
+
+        if (Strings.isNullOrEmpty(bootstrapPassword.toString())) {
+            terminal.println("'elastic' user password: " + password);
```
These will need to be parsed by another process and not seen directly from users so we can make them machine friendly ( i.e. skip single quotes)

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+            password = bootstrapPassword;
+        }
+        token = createEnrollmentToken.createKibanaEnrollmentToken(elasticUser, password);
+        fingerprint = createEnrollmentToken.getFingerprint();
```
see comment in CreateEnrollmentToken

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+        }
+        terminal.println("CA fingerprint: " + fingerprint);
+        terminal.println("Kibana enrollment token: " + token);
+        if (options.nonOptionArguments().contains("--docker")) {
```
This doesn't need to be a nonOption. We can have a proper option and it will be set by the startup script ( assuming that you have discussed this before and we have a way to determine that we are running in docker that we plan on using )

jkakavas replied:
This is not addressed @BigPandaToo

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+            bootstrapPassword;
+
+        final Environment newEnv = new Environment(settings, env.configFile());
+        Environment.assertEquivalent(newEnv, env);
```
Why do we need this assertion?

BigPandaToo replied:
We, probably don't...

jkakavas reviewed (changes requested):
Didn't get to the test class but I think it's better to look at it after some of the feedback has been addressed

BigPandaToo committed:
- Create EnrollmentToken class
- Rename couple classes for better description
- Refactor checkClusterHealth

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Pwd_initNode

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseClientAwareCommand.java:
```diff
@@ -0,0 +1,150 @@
+        HttpResponse response = null;
+        try {
+            response = client.execute("GET", clusterHealthUrl, username, password, () -> null, this::responseBuilder);
+        } catch (Exception e) {
```
ConnectException if node is not up yet. I think there may be more, if node is in the middle of starting, so I wasn't more specific

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+import java.net.URL;
+import java.util.function.Function;
+
+public class BootstrapPasswordAndEnrollmentForInitialNode extends BaseClientAwareCommand {
```
Sorry, didn't get this comment...

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
```
But it can be invoked. I don't have strong opinion on this though, but doesn't heart leaving it here

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
+            throw new UserException(ExitCodes.CONFIG, "This command is not intended for end users");
+        }
+        if (env.settings().hasValue(XPackSettings.ENROLLMENT_ENABLED.getKey()) && false ==
```
Currently it can be invoked on it's own. Again, won't fight for it...

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,199 @@
+        return password;
+    }
+
+    void ReadBootstrapPassword (Environment env, Terminal terminal) throws Exception {
```
https://github.com/elastic/elasticsearch/pull/75816#discussion_r682969333

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
```
https://github.com/elastic/elasticsearch/pull/75816#discussion_r685256471

> But it can be invoked. I don't have strong opinion on this though, but doesn't heart leaving it here

How can it be invoked when we don's ship a CLI tool ? You mean a user writing a script on their own that tries to use this as a main class ? If they go to this extent, what stops them from passing --explicitly-acknowledge-execution as a parameter while they're at it ?

jkakavas commented:
@BigPandaToo I started reviewing this as requested but then I realized that there is a big number of my comments from the previous review that haven't been addressed. Can you please address these ( or comment when you disagree ) before asking to re-review this ?

jkakavas reviewed (commented):
I will complete the review once previous comments have been addressed/discussed, to make better use of both our times

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentForInitialNode.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    void ReadBootstrapPassword (Environment env, Terminal terminal) throws Exception {
+        decryptKeyStore(keyStoreWrapper, terminal);
```
We can pass it as an argument. The method will expect the first argument to be the keystore password

jkakavas replied:
> The method will expect the first argument 

No need to use positional arguments. We can have a named parameter

BigPandaToo committed:
Addressing more PR feedback

BigPandaToo commented:
> @BigPandaToo I started reviewing this as requested but then I realized that there is a big number of my comments from the previous review that haven't been addressed. Can you please address these ( or comment when you disagree ) before asking to re-review this ?

I think, you probably looked at it in between of the two commits. Sorry, I probably re-requested the PR before I finished.

BigPandaToo committed:
Addressing more PR feedback

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+            throw new UserException(ExitCodes.USAGE, null);
+        }
+
+        final SecureString keystorePassword = new SecureString(keyStorePassword.values(options).get(0));
```
why not `keyStorePassword.value(options)` instead ? Also please don't use the deprecated SecureString ctor, call `toCharArray()` on the string instead

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+    final SecureRandom secureRandom = new SecureRandom();
+
+    // Package-private for testing
+    CommandLineHttpClient getClient(Environment env) {
```
That's merely preference, so just a nit. I wouldn't add these constructors only so that we can mock the method in tests. You can mock the object it returns instead

BigPandaToo replied:
I need real client in test to test the checkClusterHealthWithRetries

jkakavas replied:
>  need real client in test to test the checkClusterHealthWithRetries

Apologies, I don't follow this, can you elaborate ?

You use a real client in tests by doing `CommandLineHttpClient client = new CommandLineHttpClient(environment);` already. You are not using `getClient()` for that. In fact, you are using `getClient()` only with the `mock(CommandLineHttpClient)`

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (!options.has(keyStorePassword)) {
+            throw new UserException(ExitCodes.USAGE, null);
```
I thought we decided to not throw UserExceptions in this class, right?

BigPandaToo replied:
See above

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+            password = setElasticUserPassword(client, bootstrapPassword);
+            terminal.println("elastic user password: " + password);
+        }
+        final EnrollmentToken kibanaToken;
```
nit, you can move the println in the try clause so that you don't have to split the declaration above

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+        terminal.println("CA fingerprint: " + kibanaToken.getFingerprint());
+        terminal.println("Kibana enrollment token: " + kibanaToken.encode());
+        if (options.has(docker)) {
+            final EnrollmentToken nodeToken;
```
nit, you can move the println in the try clause so that you don't have to split the declaration above

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+        }
+
+        final SecureString keystorePassword = new SecureString(keyStorePassword.values(options).get(0));
+        SecureString password;
```
I think that if you first generate tokens and then set the password of the elastic user, it would allow you to simplify your code as you don't need to keep track of password here and first use boostrapPassword. You can keep using bootstrapPassword for everyhing and just print the autogenerated password of the user in the end

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+        if (ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD.exists(env.settings())) {
+            password = bootstrapPassword;
+        } else {
+            password = setElasticUserPassword(client, bootstrapPassword);
```
what is the expected behavior if we can't set the elastic user password ? Should we fail and throw as we do now, or should we attempt to proceed with generating the tokens as we do when bootstrap.password is defined ?

BigPandaToo replied:
After moving the password setting at the end, we will always generate and print out the token

jkakavas replied:
I was mostly interested in if this has been already discussed and what we _think_ we should to be doing.

> After moving the password setting at the end, we will always generate and print out the token

This is not necessary. We can select to do otherwise, if we think it makes sense

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,187 @@
+
+public class BootstrapPasswordAndEnrollmentTokenForInitialNode extends KeyStoreAwareCommand {
+    private static final String elasticUser = ElasticUser.NAME;
+    private KeyStoreWrapper keyStoreWrapper;
```
can keyStoreWrapper be a local variable as it is only used in readBootstrapPassword ?

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNodeTests.java:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    public void testGenerateNewPasswordSuccess() throws Exception {
+        final BootstrapPasswordAndEnrollmentTokenForInitialNode command_util = mock(BootstrapPasswordAndEnrollmentTokenForInitialNode
```
we don't usually name our variables like this, we stick to camelCase, so commandUtil

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNodeTests.java:
```diff
@@ -0,0 +1,318 @@
+    public void testGenerateNewPasswordSuccess() throws Exception {
+        final BootstrapPasswordAndEnrollmentTokenForInitialNode command_util = mock(BootstrapPasswordAndEnrollmentTokenForInitialNode
+            .class);
+        final CommandLineHttpClient client_util = mock(CommandLineHttpClient.class);
```
clientUtil, etc below

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNodeTests.java:
```diff
@@ -0,0 +1,318 @@
+        final EnrollmentToken token = new EnrollmentToken("DR6CzXkBDf8amV_48yYX:x3YqU_rqQwm-ESrkExcnOg",
+            "ce480d53728605674fcfd8ffb51000d8a33bf32de7c7f1e26b4d428f8a91362d", "8.0.0",
+            Arrays.asList("[192.168.0.1:9201, 172.16.254.1:9202"));
+        final BootstrapPasswordAndEnrollmentTokenForInitialNode command = new BootstrapPasswordAndEnrollmentTokenForInitialNode() {
```
we do we have both a mock (command_util) and an implementation of BootstrapPasswordAndEnrollmentTokenForInitialNode ?

BigPandaToo replied:
Yes, I want to be able to reach the real methods implementations (like checkClusterHealth) while mocking the main command.

jkakavas replied:
I'm not sure you need to do this. For instance, you don't need to do 
```
final BootstrapPasswordAndEnrollmentTokenForInitialNode command_util = mock(BootstrapPasswordAndEnrollmentTokenForInitialNode
            .class);
doReturn(client_util).when(command_util).getClient(environment);
```
, you are already doing this in 

```
@Override
protected CommandLineHttpClient getClient(Environment env) {
    return client_util;
}
```

a few lines below. You can see that if you remove the first 2 lines, the test still passes.

BigPandaToo replied:
I refactored the tests, didn't need it indeed

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNodeTests.java:
```diff
@@ -0,0 +1,318 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class BootstrapPasswordAndEnrollmentTokenForInitialNodeTests extends ESTestCase {
```
You can extend CommandTestCase which will simplify your code below

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/EnrollmentToken.java:
```diff
@@ -0,0 +1,51 @@
+        builder.field("key", apiKey);
+        builder.endObject();
+        final String jsonString = Strings.toString(builder);
+        return Base64.getUrlEncoder().encodeToString(jsonString.getBytes(StandardCharsets.UTF_8));
```
If we potentially ever may use it as an API parameter then we should keep the Url safe encoding. Just thinking; it may hurt to change it later w/o braking bwc. Also it doesn't affect any decoding process unlike cert encoding.

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,199 @@
+        try {
+            client.checkClusterHealthWithRetriesWaitingForCluster(elasticUser, credentialsPassword, 5);
+        } catch (Exception e) {
+            throw new UserException(ExitCodes.UNAVAILABLE, null);
```
As we discussed in Slack, it's ok as long as we don't show it stdout

BigPandaToo committed:
Addressing more PR feedback

BigPandaToo commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

BigPandaToo committed:
Fixing conflict

BigPandaToo commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

BigPandaToo committed:
Fixing conflict

BigPandaToo commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

BigPandaToo committed:
Merge branch 'master' into Pwd_initNode

BigPandaToo committed:
Fixing conflict

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java:
```diff
@@ -195,42 +193,29 @@ private void ensureFileRealmEnabled(Settings settings) throws Exception {
-                Thread.sleep(1000);
-                retries -= 1;
-                checkClusterHealthWithRetries(env, terminal, username, password, retries, force);
+            if (e instanceof ElasticsearchStatusException &&
```
This doesn't work now, because `checkClusterHealthWithRetriesWaitingForCluster` wraps exception in `IllegalStateException`s.

I'm starting to regret my suggestion to have a common approach for the cluster health polling.
My suggestion now is to do whatever works that follows the following requirements:

For `BaseRunAsSuperUserCommand`:
* fail on connection exception (node is down)
* retry on HTTP_UNAUTHORIZED and HTTP_FORBIDDEN

For `BootstrapPasswordAndEnrollmentTokenForInitialNode`:
* retry on connection exception
* fail on any Exception

I think it's better that these methods be at the `CommandLineHttpClient` level, to aid with discoverability, and that they reuse code, but I'm not going to gripe about it.

BigPandaToo replied:
`checkClusterHealthWithRetriesWaitingForCluster` is in `CommandLineHttpClient`
`checkClusterHealthWithRetries` is in `BaseRunAsSuperuserCommand` because it is superuser aware command

`checkClusterHealthWithRetriesWaitingForCluster` throws `ElasticsearchStatusException` in case of HTTP_UNAUTHORIZED and HTTP_FORBIDDEN and pass the specific http error code to `checkClusterHealthWithRetries` so it can retry in this case

jkakavas replied:
> I'm starting to regret my suggestion to have a common approach for the cluster health polling.

++ :/ I think we gave it a try and it doesn't work ( I don't like exception type based flow control ) . I'd suggest we leave the BaseRunAsSuperUserCommand check health method as it was and implement  something for BootstrapPasswordAndEnrollmentTokenForInitialNode that makes sense for it since these two have slightly different requirements. I'd leave the first in the BaseRunAsSuperUserCommand as it is very specific to it, but I don't have strong opinions on where to place the second checkclusterhealth method. 

WDYT Lyudmila ?

BigPandaToo replied:
I tend to agree with both of you. I think the current implementation would work, but it doesn't look intuitive and overall fragil.
I would leave the one waiting for cluster in `CommandLineHttpClient` as I think it is generic enough to be used by other commands

albertzaharovits reviewed (changes requested):
I did a through review.
It is close.
Please ping me when you've addressed it all and I'll approve.

BigPandaToo committed:
Simplifying tests/ Fixing tests

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,172 @@
+            terminal.println("CA fingerprint: " + kibanaToken.getFingerprint());
+        } catch (Exception e) {
+            throw new UserException(ExitCodes.UNAVAILABLE, null);
+        }
```
Agree on having one block. But what do you mean by swallowing exceptions? We don't swallow them, just don't have custom messages

albertzaharovits replied:
> what do you mean by swallowing exceptions? We don't swallow them, just don't have custom messages
```
        } catch (Exception e) {
            throw new UserException(ExitCodes.UNAVAILABLE, null);
        }
```
 -> 
 ```
        } catch (Exception e) {
            throw new UserException(ExitCodes.UNAVAILABLE, "", e);
        }
```

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,172 @@
+    private final CheckedFunction<Environment, EnrollmentTokenGenerator, Exception> createEnrollmentTokenFunction;
+    private final Function<Environment, CommandLineHttpClient> clientFunction;
+    private final CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction;
+    private final OptionSpec<Void> docker;
```
Not sure about that... Isn't it explanatory enough? Same as if you provide a "force" option

albertzaharovits replied:
Okay, if the option is named `include-node-enrollment-token` rather than `node-enrollment-token`, Void is fine.
Thanks for considering it!

BigPandaToo committed:
Addressing PR feedback

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Pwd_initNode

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/EnrollmentToken.java:
```diff
@@ -0,0 +1,64 @@
+     * @param apiKey         API Key credential in the form apiKeyId:ApiKeySecret to be used for enroll calls
+     * @param fingerprint    hex encoded SHA256 fingerprint of the HTTP CA cert
+     * @param version        node version number
+     * @param boundAddress   IP Addresses and port numbers for the interface where the Elasticsearch node is listening on
```
```suggestion
     * @param boundAddress   IP Addresses and port numbers for the interfaces where the Elasticsearch node is listening on
```

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/EnrollmentToken.java:
```diff
@@ -0,0 +1,64 @@
+        return Strings.toString(builder);
+    }
+
+    public String encode() throws Exception {
```
nit suggestion : `getEnoded()`

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,154 @@
+        this.clientFunction = clientFunction;
+        this.keyStoreFunction = keyStoreFunction;
+        this.createEnrollmentTokenFunction = createEnrollmentTokenFunction;
+        docker = parser.accepts("include-node-enrollment-token", "determine that we are running in docker");
```
let's remove docker references from here. For the same reasons you changed it to `include-node-enrollment-token`

jkakavas reviewed (commented):
Looking good Lyudmila, I added a couple of comments and a few nit suggestions. Happy to approve once comments are resolved

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/CommandLineHttpClient.java:
```diff
@@ -162,4 +172,84 @@ public String getDefaultURL() {
+    /**
+     * If cluster is not up yet (connection refused or cluster still Red), we will retry @retries number of times
+     */
+    public void checkClusterHealthWithRetriesWaitingForCluster(String username, SecureString password, int retries, boolean force)
```
Makes sense

BigPandaToo committed:
Moving checkForClusterHealth back to `BaseRunAsSuperuserCommand`;
Addressing more feedback

BigPandaToo committed:
Fixing tests

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/CommandLineHttpClient.java:
```diff
@@ -162,4 +169,84 @@ public String getDefaultURL() {
+            }
+        }
+        final int responseStatus = response.getHttpStatus();
+        if (responseStatus != HttpURLConnection.HTTP_OK) {
```
Technically I would say that 503 is also a condition (in addition to connection problems) that should be retryable.
Cluster health can return it when there's no master (it does some retrying, it's unlikely, but possible).

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,154 @@
+        final SecureString bootstrapPassword = ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD.get(secureSettings);
+        try {
+            String output;
+            client.checkClusterHealthWithRetriesWaitingForCluster(ElasticUser.NAME, bootstrapPassword, 5);
```
5 seconds for a node to come up is a bit on the low side. I would put it closer to 15.

albertzaharovits reviewed (approved):
Two more issues on the cluster health retry logic. Please address.
Besides that LGTM.

Thanks for the effort.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,154 @@
+        final SecureString keystorePassword = new SecureString(keyStorePassword);
+        final CommandLineHttpClient client = clientFunction.apply(env);
+        final EnrollmentTokenGenerator enrollmentTokenGenerator = createEnrollmentTokenFunction.apply(env);
+        final Settings secureSettings = readSecureSettings(env, keystorePassword);
```
This _needs_ to return an Environment (which will contain the secure settings from the keystore ) and you _need_ to pass this `Environment` to 
```
final CommandLineHttpClient client = clientFunction.apply(env);
final EnrollmentTokenGenerator enrollmentTokenGenerator = createEnrollmentTokenFunction.apply(env);
```

so that these classes can read key material from the TLS keystores ( that are password protected and their passwords are stored in the secure settings ).

Apologies I missed it yesterday when I made the comment that this can return Settings instead, we weren't using the returned environment below and it didn't ring any bells

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,154 @@
+            response = client.execute("POST", passwordSetUrl, ElasticUser.NAME, bootstrapPassword,
+                () -> {
+                    XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+                    xContentBuilder.startObject().field("password", password).endObject();
```
You can't pass a `SecureString` as the value in `field()`, you need to use the char array here and wrap it in a SecureString later or call `toString()` on the `SecureString` you already have

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/CommandLineHttpClient.java:
```diff
@@ -162,4 +169,84 @@ public String getDefaultURL() {
+                throw new IllegalStateException(
+                    "Failed to determine the health of the cluster. Cluster health API did not return a status value."
+                );
+            } else if ("red".equalsIgnoreCase(clusterStatus)) {
```
I like this! I have changed to wait for yellow for default 30s instead of reiterating here.

BigPandaToo replied:
Reiterating only waiting for connection

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java:
```diff
@@ -0,0 +1,154 @@
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        final char[] keyStorePassword;
```
let's not have two variables of different type with _almost_ the same name if we don't need to. You can simply have

final SecureString keystorePassword; 
try {
    keystorePassword = new SecureString(terminal.readSecret(""));
} catch (Exception e) {
    throw new UserException(ExitCodes.USAGE, null);
}

jkakavas reviewed (approved):
LGTM provided you address Albert's and my final comments before merging. Thanks for the iterations Lyudmila!

p.s. Lack of QA testing is biting us, both the issues I commented on in my last review could not be found with unit testing. We need to discuss relevant priorities and find a useful balance between feature mvp and useful/necessary testing.

BigPandaToo committed:
Some changes to `checkClusterHealthWithRetriesWaitingForCluster':
- Reiterate if http status 503 terturned
- Use 'wait for yellow' option instead of reiterating
- Increase to 15 retrys

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Pwd_initNode

BigPandaToo committed:
Addressing more PR feedback
